### PR TITLE
Simplify fetch actions api

### PR DESCRIPTION
**Pull request by samwinslow**
Essentially the logic of https://github.com/PostHog/growth-accounting/pull/2/files, but in a plugin.

### CODE REVIEW

**Comment by samwinslow** [comment]
Not yet tested, see https://posthog.slack.com/archives/C02E3BKC78F/p1637196847263900

File: index.ts [hunk 1]
```diff
@@ -1,273 +1,272 @@
...
-        return null
+    // Only fetch again if in the steady state
+    const shouldFetchEvents = eventStorage.every(({ loadingState }) => loadingState === null)
+    if (shouldFetchEvents) {
+        const after = await storage.get('last_invoked_at')
+        if (!isValidDate(after)) {
+            throw new Error(`Missing or invalid last_invoked_at: "${after}"`)
+        }
+        for (const actionId in eventsByAction) {
+            await jobs.getEventsForAction({
```
Comment (kpthatsme) [inline index.ts:224]: Why aren't we just doing these calls sequentially and storing the result before moving to the next step?

Seems like that would be simpler unless I'm missing something

File: index.ts [hunk 2]
```diff
@@ -1,273 +1,272 @@
...
     }
-    const conversion = conversionDefinitions.find(({ eventDetails }) =>
-        eventMatchesDefinition(event, eventDetails)
-    )
-    if (conversion) {
-        if (!event.properties.gclid) {
-            return null
+
+    // If all events are loaded...
+    const eventsLoaded = eventStorage.every(({ loadingState }) => loadingState === 'loaded')
```
Comment (kpthatsme) [inline index.ts:232]: I don't think this works because there's no mechanism to wake up when these are completed - so the job just won't complete.

**Review by kpthatsme — COMMENTED** [review]
@samwinslow I left some feedback -

Generally I think you're definitely going down the right path, but my overall sense is that we can simplify some of this- specifically the recursion and the attempted use of shared storage.

I think things would be a lot easier to reason about + test (and maybe performance) if you approach this more sequentially than the way you've done it now. 

All of the APIs we're using to collect data from PostHog naturally support pagination in an iterative way too - just continue hitting the next url until it no longer comes back.

Here's an example of what I'm thinking:

```
action_conversions = load_all_conversions(last_run_ts)
distinct_ids_to_gclid = {}

for event in  action_conversion:
   gclid = gclid_from_event(event)
   distinct_id = event.distinct_id
   if gclid: 
     write this conversion to zapier
   if there is no gclid:
     if distinct_id not in distinct_ids_to_gclid: ## need to load the user
          distinct_ids_to_gclid[distinct_id] = load_user_props(distinct_id) ## if you want to save memory only save the gclid prop in the cache
     gclid = distinct_ids_to_gclid[distinct_id].get(gclid)
     if gclid:
        write this conversion to zapier
  
```

Did you think about trying an approach like this? Is there somewhere it didn't work?

I could see one concern being time to execute – but I don't think that's a problem once the since the Plugin scheduler is smart enough to know not to start the next run until _after_ the previous run finished.

**Comment by samwinslow** [comment]
Hey @kpthatsme, I can dive into the specifics of why sequentially making lots of API calls will not work, but there's now a PR #4 that will replace all of this. It operates on each action, rather than worrying about batches.

For context the reason calling everything sequentially would fail is that the plugin server **will error** on:

- any Promise that takes 30s or more to resolve/reject
- any loop that is run for more than 30s.

Jobs are a way around the problems of long-running `await`ed tasks, and they have several advantages:

- not doing too much work on the main thread / event loop
- ability for the plugin-server to queue, retry, or reject jobs as needed

When we `await` we're forcing async tasks to behave as if they're synchronous and not taking advantage of the plugin server's inherently multi-threaded structure. I think https://github.com/samwinslow/mailchimp-campaigns-plugin is an example of jobs done right, or at least better than what's here — in any case though this is made irrelevant by the addition of `onAction`.

**Comment by kpthatsme** [comment]
> For context the reason calling everything sequentially would fail is that the plugin server **will error** on:
> 
> * any Promise that takes 30s or more to resolve/reject

We shouldn't need to run any single requests that take that long..are you running into that? If we are we should process smaller batches

> * any loop that is run for more than 30s.

Not sure if I follow this point as much I'm not saying we're doing this, but It's possible to do work in a `runEveryMinute` that takes hours to complete - all other runs of this job will be queued up behind it though.

Will follow up on the other PR.

**Comment by samwinslow** [comment]
> We shouldn't need to run any single requests that take that long

Correct, and this PR doesn't have any requests which break that rule.

From the pseudo-code example you gave, a for loop within one function that calls an API again and again is liable to run for longer than 30s, and be killed.

> it's possible to do work in a runEveryMinute

Yep! And this PR does that. That's why line 232 would work -- there are several high-level fetching "modes" of the program,

1. Fetch Events
2. Fetch Persons, if events are completed
3. Post to Zapier, if events and persons are completed
4. All fetch tasks done -> clear global state and reset.

runEveryMinute will initiate the next job in the sequence, or do nothing for the jobs that are still running, so on through the sequence above.

Each job calls itself again and again until it has nothing left to fetch. This looks like recursion, but doesn't cause 30 second timeouts because `await someJob().runNow()` resolves immediately.

image

It works but is a bit hard to reason around.

**Pull request closed without merging**